dotcom-rendering/src/web/components/DynamicSlowMPU.tsx

```diff
@@ -0,0 +1,382 @@
+		>
+			<FrontCard
+				trail={cards[1]}
+				containerPalette={containerPalette}
+				showAge={showAge}
+				trailText={
+					cards[1]?.supportingContent &&
+					cards[1].supportingContent.length > 0
+						? undefined
+						: cards[1].trailText
+				}
+				supportingContent={
+					cards[1].trailText ? undefined : cards[1].supportingContent
+				}
+			/>
+		</LI>
+		<LI
+			percentage="25%"
+			padSides={true}
+			showDivider={true}
+			showTopMarginWhenStacked={true}
+		>
+			<FrontCard
+				trail={cards[2]}
+				containerPalette={containerPalette}
+				showAge={showAge}
+				trailText={
+					cards[2]?.supportingContent &&
+					cards[2].supportingContent.length > 0
+						? undefined
+						: cards[2].trailText
+				}
+				supportingContent={
+					cards[2].trailText ? undefined : cards[2].supportingContent
+				}
+			/>
+		</LI>
+	</UL>
+);
+
+/**
+ * DynamicSlowMPU
+ *
+ * This container only allows big and standard cards (from groupedTrails)
+ *
+ * Layout is dynamic deopending on the number of big cards. You're only
+ * allowed to have 2 or 3 big cards. If you pass 1 a standrd card will
+ * get promoted to make two bigs. If you pass more than 3 bigs then the
+ * extras will get demoted to standard.
+ *
+ * If you pass no bigs at all the top slice will not render and a special
+ * 3 column layout is used for the remaining slice.
+ *
+ */
+export const DynamicSlowMPU = ({
+	groupedTrails,
+	containerPalette,
+	showAge,
+}: Props) => {
+	let bigCards = groupedTrails.big;
+	let standardCards = groupedTrails.standard;
+
+	if (bigCards.length === 0) {
+		// If there are no big cards at all we get just one, special, slice:
+		return (
+			<Card33_ColumnOfThree33_Ad33
+				cards={standardCards}
+				containerPalette={containerPalette}
+				showAge={showAge}
+			/>
+		);
+	}
+
+	if (bigCards.length === 1) {
+		// If there is only one big item a standard item is promoted such that
+		// there is always at least two
+		const promoted = standardCards.slice(0, 1);
+		bigCards = [...bigCards, ...promoted];
+		standardCards = standardCards.slice(1);
+	}
+
+	if (bigCards.length > 3) {
+		// If there are more than three big cards, these extra bigs are demoted to
+		// standard.
+		const demoted = bigCards.slice(2);
+		standardCards = [...demoted, ...standardCards];
+	}
+
+	if (bigCards.length === 2 && bigCards[0].isBoosted) {
+		return (
+			<>
+				<Card75_Card25
+					cards={bigCards}
+					containerPalette={containerPalette}
+					showAge={showAge}
+				/>
+				<ColumnOfThree50_Ad50
+					cards={standardCards}
+					containerPalette={containerPalette}
+					showAge={showAge}
+				/>
+			</>
+		);
+	}
+
+	if (bigCards.length === 2) {
+		return (
+			<>
+				<Card50_Card50
+					cards={bigCards}
+					containerPalette={containerPalette}
+					showAge={showAge}
+				/>
+				<ColumnOfThree50_Ad50
+					cards={standardCards}
+					containerPalette={containerPalette}
+					showAge={showAge}
+				/>
+			</>
+		);
+	}
+
+	// Else bigCards.length === 3
```

Comment:
Logically, bigCards has a length of three here but typescript isn't following along so I'm defaulting to this outcome.

Reply:
I could definitely be wrong here, but: is typescript complaining because it's possible that `promoted` in ln.322 is just an empty array, in which case at this point `bigCards.length === 1`?

Reply:
I'm finding the way that length checks are done both before and after the array is mutated seems to be making it a bit more difficult to follow the logic here. Have you considered any alternative ways of implementing this logic? One option that comes to mind is to break into two steps, something like `decidePromotion` and then `decideContainer` (though not sure that they'd need to be actually separate functions as such).

I'm happy to try and sketch what I have in mind in a bit more detail. If you think it sounds promising though then lmk and I'll try to describe it a bit clearer than I've done here

Reply:
> I could definitely be wrong here, but: is typescript complaining because it's possible that `promoted` in ln.322 is just an empty array, in which case at this point `bigCards.length === 1`?

The fact that each branch is wrapped in a conditional means typescript will complain that there is not a certain return path, even if we know that it is sure that one of the branches will be followed.

I think this is fine. We're saying that the default is to render `Card50_Card25_Card25`.

Reply:
> I'm finding the way that length checks are done both before and after the array is mutated seems to be making it a bit more difficult to follow the logic here. Have you considered any alternative ways of implementing this logic? One option that comes to mind is to break into two steps, something like `decidePromotion` and then `decideContainer` (though not sure that they'd need to be actually separate functions as such).

You are 100% spot on with this. I implemented your suggestion and the code reads so much better. Thanks!

Reply:
glad to hear it was a helpful suggestion :) and yeah fwiw I find it a lot easier to follow the logic now 👍